src/aoutils.cpp

```diff
@@ -0,0 +1,81 @@
+#include "aoutils.h"
+
+#include <QDebug>
+#include <QMap>
+#include <QPair>
+#include <QRegularExpression>
+#include <QRegularExpressionMatch>
+#include <QSettings>
+
+void AOUtils::migrateEffects(QSettings &p_effects_ini)
+{
+  qDebug() << "Migrating effects from file:" << p_effects_ini.fileName();
+
+  QMap<QString, QString> l_effect_map;
+  {
+    const QStringList l_key_list = p_effects_ini.childKeys();
+    for (const QString &i_key : l_key_list)
+    {
+      l_effect_map.insert(i_key, p_effects_ini.value(i_key).toString());
+    }
+
+    p_effects_ini.clear();
+    p_effects_ini.sync();
+  }
+  // update revision
+  p_effects_ini.beginGroup("version");
+  p_effects_ini.setValue("major", "2");
+  p_effects_ini.endGroup();
+
+  const QStringList l_property_list{
+    "sound",
+    "scaling",
+    "stretch",
+    "ignore_offset",
+    "under_chatbox",
+  };
+
+  const QMap<QString, QPair<QString, QString>> l_property_replacement_list{
+    {"under_chatbox", {"layer", "character"}},
+  };
+
+  QStringList l_key_list;
+  const QRegularExpression l_regex(QStringLiteral("(\\w+)_(%1)$").arg(l_property_list.join("|")));
+  for (auto i = l_effect_map.begin(); i != l_effect_map.end(); i++)
+  {
+    if (l_regex.match(i.key()).hasMatch())
+    {
+      continue;
+    }
+    l_key_list.append(i.key());
+  }
+
+  int i = 0;
+  for (const QString &i_effect_key : qAsConst(l_key_list))
+  {
+    p_effects_ini.beginGroup(QString::number(i++));
+    p_effects_ini.setValue("name", i_effect_key);
```

Comment:
p and i are terrible variable names

Reply:
They are mostly there to clarify the lifetime of these variables to be local to the function and that the effects_ini is a pointer.